sdk/samples/iot/paho_iot_pnp_component_sample.c

```diff
@@ -0,0 +1,626 @@
+  az_result rc = az_iot_hub_client_properties_document_get_publish_topic(
+      &hub_client,
+      pnp_mqtt_get_request_id(),
+      publish_message.topic,
+      publish_message.topic_length,
+      NULL);
+  IOT_SAMPLE_EXIT_IF_AZ_FAILED(rc, "Failed to get the property document topic");
+
+  // Publish the property document request.
+  publish_mqtt_message(
+      mqtt_client, publish_message.topic, AZ_SPAN_EMPTY, IOT_SAMPLE_MQTT_PUBLISH_QOS);
+  IOT_SAMPLE_LOG(" "); // Formatting
+}
+
+// receive_messages_and_send_telemetry_loop will loop to check if there are incoming MQTT
+// messages, waiting up to MQTT_TIMEOUT_RECEIVE_MS.  It will also send a telemetry message
+// every time through the loop.
+static void receive_messages_and_send_telemetry_loop(void)
+{
+  // Continue to receive commands or device property messages while device is operational.
+  while (is_device_operational)
+  {
+    // Send telemetry messages. No response requested from server.
+    send_telemetry_messages();
+    IOT_SAMPLE_LOG(" "); // Formatting.
+
+    // Wait for any server-initiated messages.
+    receive_mqtt_message();
+  }
+}
+
+// disconnect_mqtt_client_from_iot_hub disconnects and destroys the underlying MQTT connection and
+// Paho handle.
+static void disconnect_mqtt_client_from_iot_hub(void)
+{
+  int rc = MQTTClient_disconnect(mqtt_client, MQTT_TIMEOUT_DISCONNECT_MS);
+  if (rc != MQTTCLIENT_SUCCESS)
+  {
+    IOT_SAMPLE_LOG_ERROR("Failed to disconnect MQTT client: MQTTClient return code %d.", rc);
+    exit(rc);
+  }
+
+  MQTTClient_destroy(&mqtt_client);
+}
+
+// receive_mqtt_message checks for incoming messages from Azure IoT Hub.
+static void receive_mqtt_message(void)
+{
+  char* topic = NULL;
+  int topic_len = 0;
+  MQTTClient_message* receive_message = NULL;
+  static int8_t timeout_counter;
+
+  IOT_SAMPLE_LOG("Waiting for command request or device property message.\n");
+
+  // MQTTCLIENT_SUCCESS or MQTTCLIENT_TOPICNAME_TRUNCATED if a message is received.
+  // MQTTCLIENT_SUCCESS can also indicate that the timeout expired, in which case message is NULL.
+  // MQTTCLIENT_TOPICNAME_TRUNCATED if the topic contains embedded NULL characters.
+  // An error code is returned if there was a problem trying to receive a message.
+  int rc = MQTTClient_receive(
+      mqtt_client, &topic, &topic_len, &receive_message, MQTT_TIMEOUT_RECEIVE_MS);
+  if ((rc != MQTTCLIENT_SUCCESS) && (rc != MQTTCLIENT_TOPICNAME_TRUNCATED))
+  {
+    IOT_SAMPLE_LOG_ERROR("Failed to receive message: MQTTClient return code %d.", rc);
+    exit(rc);
+  }
+  else if (receive_message == NULL)
+  {
+    // Allow up to TIMEOUT_MQTT_RECEIVE_MAX_COUNT before disconnecting.
+    if (++timeout_counter >= MQTT_TIMEOUT_RECEIVE_MAX_MESSAGE_COUNT)
+    {
+      IOT_SAMPLE_LOG(
+          "Receive message timeout expiration count of %d reached.",
+          MQTT_TIMEOUT_RECEIVE_MAX_MESSAGE_COUNT);
+      is_device_operational = false;
+    }
```

Comment:
The ;tl-dr here @momuno is could you please start an offline discussion with folks including me that need to weigh in on fixing other samples.  The sample in `receive_c2d_messages` seems to be doing same, e.g., and I want PnP to be inline with everything else.

I'm fine changing code to be more resistant in case Paho would ever change, though I think if Paho did they'd cause leaks for other stuff.

I do think the code is technically correct as is.  I took a look at Paho (https://github.com/eclipse/paho.mqtt.c/blob/master/src/MQTTClient.c) and `MQTTClient_receive` and the call that fills out structures is `MQTTClient_deliverMessage` and is only filled if there's data.  `MQTTClient_deliverMessage` fills both topic and message in the same time, so presumably we can't get a NULL message but a topic that'd need to be freed.

Looking to where the structure gets filled out brings me to `Protocol_processPublication` and it setting a non-NULL `msg`, so even if there's topic truncation we'd still get a non-NULL message up at our sample layer which would trigger freeing the topic.  I'm not seeing any other way for a message to be put into the list here from my review.

Reply:
Ideally their docs should be updated to reflect that :)

Reply:
Sorry didn't mean to dogpile I was mid review when John replied. Didn't see.

Reply:
No worries and thanks for extra eyes on this, Dane.